### DISCUSSION TITLE
AVI movie data altered after resaving file with movie.save

### DISCUSSION THREAD

**Discussion by rpfisher9** [flatironinstitute/CaImAn#1178] [Q&A] [answered]
Hi!
When analyzing our .avi movie files from 1p calcium imaging in CaImAn (version 1.9.15), we noticed videos resaved as a new file with  movie.save have a different 3D movie array than the original avi file (see code screenshot below). There were no other changes made to the videos other than loading and taking the average. Has anyone else noticed modified data from videos after saving as a new file or know why this may be happening? 
![resaving_avi](https://github.com/flatironinstitute/CaImAn/assets/133471152/0af9879a-8a11-47ec-9640-d0edcf6505c8)

**Comment by EricThomson** [comment]
I would inspect the movie and make sure it still looks reasonable. 

The load function converts to float, and also does an `rgb2grey` operation on it that can also change the absolute values (this latter operation shouldn't change the relative values if your original is grayscale). We change the data type because there is later lots of floating point arithmetic done on the arrays in the movie.

**Reply by rpfisher9** [reply]
Hi, 
Thank you for the quick response! Our input AVI (shape [1000,608,608]) files are in greyscale and the videos still look reasonable. However, the change in the movie values does not appear to be in load. Below is a graph of a column from a Miniscope frame graph of our original AVI file (in red) and the same column of data from a video that has been saved and reloaded (in blue). Both were loaded in the same manner, so I dont think specifics of the load method explains why the video that was saved and opened has movie values that are so much higher. The graph also appears to artificially cut off at 255, causing the shape of the blue resaved movie to distort.  

![original_v_resaved](https://github.com/flatironinstitute/CaImAn/assets/133471152/83066993-058d-4267-82d7-d9e905612973)
![code_for_graph](https://github.com/flatironinstitute/CaImAn/assets/133471152/c8c55c94-2777-4300-b7d6-467132b06620)

**Reply by rpfisher9** [reply]
Hi, following up with some more information in case it helps! 
For the above graph, Ive attached reference images with an example column of data shown by the white rectangle. Visually, it is  apparent that the frame from the original movie is quite different from the resaved. Shifts in the data's absolute brightness values are not a huge concern, but if the data is being clipped (like it appears to be in the blue trace) would be a problem. 
![sleep_original](https://github.com/flatironinstitute/CaImAn/assets/133471152/596ad043-70dc-43ad-a194-4e1fa512b76d)
![sleep_resaved](https://github.com/flatironinstitute/CaImAn/assets/133471152/0e366738-38b1-4d36-87df-12ec4af464bf)

**Reply by rpfisher9** [reply]
Additionally, Below is a graph of an original (red) and resaved (blue) video from a test recording that had the miniscope recording with no GRIN lens or test subject. The only code run on these videos is load and save (with their default settings), like above. Any thoughts you have would be appreciated! 
![Figure_2](https://github.com/flatironinstitute/CaImAn/assets/133471152/95ec4cbe-f8cb-4a0e-b514-b9140819ce19)

**Comment by EricThomson** [comment]
This is strange. What operating system are you in?

If not Windows, if it isn't too much trouble, could you share a small movie (e.g., a couple of hundred frames) and small self-contained code file that you used to recreate these issues?

**Reply by rpfisher9** [reply]
We are using Linux Mint 21.1 Cinnamon OS and Spyder as the environment.
Here is a dropbox link to the two short videos that I have been using and a text file with the few lines of code that I used to generate the plot. 
https://www.dropbox.com/scl/fo/ekw5lvtyy606gvjac7hvj/h?rlkey=niqyh3qn07z1r46p5ulx7ezxi&dl=0
[alteredMovieAnalysis.txt](https://github.com/flatironinstitute/CaImAn/files/12785313/alteredMovieAnalysis.txt)

Thank you!

**Reply by EricThomson** [reply]
Thanks! I should be able to look at this later this week.

**Reply by rpfisher9** [reply]
Hi! Just checking in to see if you have any thoughts on what may be going on here. Thank you in advance

**Reply by EricThomson** [reply]
Sorry I am at a conference -- I will check this out soon!

**Reply by EricThomson** [reply, marked as answer]
Ok I tried to reproduce and found a bug in Windows save for avi 😭 So I'll need to work on that. 

It looks like the default for avi, though, looking on the code is to save with lots of  scaling operations that makes it better for viewing more than for processing. If you want to stop that and keep things the same (again, without actually testing it), try setting the `q_min` and `q_max` arguments to `None`, and it will turn this off. Then it should give the same outputs as inputs -- if you don't do that it will try to scale the input to the full dynamic range, which is what we are seeing in your example so it looks nicer for presentations etc. 

This is because saving avi isn't really ever used on the back end for processing in caiman it is looked at as something you might want to do to save for a presentation, so things are scaled to look nice.

If your pipeline relies on saving using caiman's `save()`, I would consider saving to a different format than avi because we have been uncovering issues lately with avi and it is not one of our favorite or most heavily vetted formats. :smile: So if you are already going to resave your data anyway, I'd consider tiff, hdf5.

**Reply by rpfisher9** [reply]
Hi! Yes it appears that was the default that was changing our movie data, thank you for the help! The pipeline does not require cm.save, I was attempting to flip an avi video because the miniscope was accidentally placed in upside down during a recording and I am hoping to avoid having to rerecord that experiment.
